File: ruby/anagram/anagram.rb
module StringHelpers
  refine String do
    def sort
      self.chars.sort.join
    end
  end
end

class Anagram
  using StringHelpers

  def initialize(word)
    @word = word.downcase
  end

  def match(words)
    words
      .select { |other_word| anagram?(other_word.downcase) } 
  end

  private
    attr_reader :word

    def anagram?(other_word)
      return false if other_word == word

      return false unless word.size == other_word.size

      other_word.sort == word.sort
    end
end

File: typescript/anagram/anagram.ts
class Anagram {
  private word: string

  constructor(word: string) {
    this.word = word.toLowerCase()
  }

  static sortLetters(word: string) {
    return word.toLowerCase().split('').sort().join('')
  }

  matches(...words: string[]) {
    return words.filter(this.isAnagram.bind(this))
  }

  isAnagram(word: string) {
    if (this.exactSameWord(word)) {
      return false
    }

    return Anagram.sortLetters(word) === Anagram.sortLetters(this.word)
  }

  exactSameWord(word: string) {
    return this.word === word.toLowerCase()
  }
}

export default Anagram